File: libmuscle/python/libmuscle/mmp_client.py
import dataclasses
from pathlib import Path
from random import uniform
from threading import Lock
from time import perf_counter, sleep
from typing import Any, Dict, Iterable, List, Optional, Tuple

import msgpack
from ymmsl import (
        Conduit, Operator, Port, Reference, Settings, Checkpoints,
        CheckpointRule, CheckpointRangeRule, CheckpointAtRule)

import libmuscle
from libmuscle.mcp.protocol import RequestType, ResponseType
from libmuscle.mcp.tcp_transport_client import TcpTransportClient
from libmuscle.profiling import ProfileEvent
from libmuscle.logging import LogMessage
from libmuscle.snapshot import SnapshotMetadata


PEER_TIMEOUT = 600
PEER_INTERVAL_MIN = 5.0
PEER_INTERVAL_MAX = 10.0

_CheckpointInfoType = Tuple[
        float, Checkpoints, Optional[Path], Optional[Path]]


def encode_operator(op: Operator) -> str:
    """Convert an Operator to a MsgPack-compatible value."""
    return op.name


def encode_port(port: Port) -> List[str]:
    """Convert a Port to a MsgPack-compatible value."""
    return [str(port.name), encode_operator(port.operator)]


def encode_profile_event(event: ProfileEvent) -> Any:
    """Converts a ProfileEvent to a list.

    Args:
        event: A profile event

    Returns:
        A list with its attributes, for MMP serialisation.
    """
    if event.start_time is None or event.stop_time is None:
        raise RuntimeError(
                'Incomplete ProfileEvent sent. This is a bug, please'
                ' report it.')

    encoded_port = encode_port(event.port) if event.port else None
    return [
            event.event_type.value,
            event.start_time.nanoseconds, event.stop_time.nanoseconds,
            encoded_port, event.port_length, event.slot,
            event.message_number, event.message_size, event.message_timestamp]


def decode_checkpoint_rule(rule: Dict[str, Any]) -> CheckpointRule:
    """Decode a checkpoint rule from a MsgPack-compatible value."""
    if rule.keys() == {'at'}:
        return CheckpointAtRule(**rule)
    if rule.keys() == {'start', 'stop', 'every'}:
        return CheckpointRangeRule(**rule)
    raise ValueError(f'Cannot convert {rule} to a checkpoint rule.')


def decode_checkpoint_info(
        elapsed_time: float,
        checkpoints_dict: Dict[str, Any],
        resume: Optional[str],
        snapshot_dir: Optional[str]
        ) -> _CheckpointInfoType:
    """Decode checkpoint info from a MsgPack-compatible value.

    Args:
        elapsed_time: current elapsed time according to the manager
        checkpoints_dict: checkpoint definitions from the MsgPack
        resume: path to the snapshot we should resume from, if any
        snapshot_dir: path to the directory to store new snapshots in

    Returns:
        elapsed_time: current elapsed time according to the manager
        checkpoints: checkpoint configuration
        resume: path to the snapshot we should resume from, if any
        snapshot_dir: path to the directory to store new snapshots in
    """
    checkpoints = Checkpoints(
            at_end=checkpoints_dict["at_end"],
            wallclock_time=[decode_checkpoint_rule(rule)
                            for rule in checkpoints_dict["wallclock_time"]],
            simulation_time=[decode_checkpoint_rule(rule)
                             for rule in checkpoints_dict["simulation_time"]])
    resume_path = None if resume is None else Path(resume)
    snapshot_path = None if snapshot_dir is None else Path(snapshot_dir)
    return (elapsed_time, checkpoints, resume_path, snapshot_path)


class MMPClient():
    """The client for the MUSCLE Manager Protocol.

    This class connects to the Manager and communicates with it on
    behalf of the rest of libmuscle.

    It manages the connection, and converts between our native types
    and the gRPC generated types.

    Communication is protected by an internal lock, so this class can
    be called simultaneously from different threads.
    """
    def __init__(self, instance_id: Reference, location: str) -> None:
        """Create an MMPClient

        Args:
            location: A connection string of the form hostname:port
        """
        self._instance_id = instance_id
        self._transport_client = TcpTransportClient(location)
        self._mutex = Lock()

    def close(self) -> None:
        """Close the connection

        This closes the connection. After this no other member
        functions can be called.
        """
        self._transport_client.close()

    def submit_log_message(self, message: LogMessage) -> None:
        """Send a log message to the manager.

        Args:
            message: The message to send.
        """
        request = [
                RequestType.SUBMIT_LOG_MESSAGE.value,
                message.instance_id, message.timestamp.seconds,
                message.level.value, message.text]
        self._call_manager(request)

    def submit_profile_events(self, events: Iterable[ProfileEvent]) -> None:
        """Sends profiling events to the manager.

        Args:
            events: The events to send.
        """
        request = [
                RequestType.SUBMIT_PROFILE_EVENTS.value,
                str(self._instance_id),
                [encode_profile_event(e) for e in events]]
        self._call_manager(request)

    def submit_snapshot_metadata(
            self, snapshot_metadata: SnapshotMetadata) -> None:
        """Send snapshot metadata to the manager.

        Args:
            snapshot_metadata: Snapshot metadata to supply to the manager.
        """
        request = [
                RequestType.SUBMIT_SNAPSHOT.value,
                str(self._instance_id),
                dataclasses.asdict(snapshot_metadata)]
        self._call_manager(request)

    def get_settings(self) -> Settings:
        """Get the central settings from the manager.

        Returns:
            The requested settings.
        """
        request = [RequestType.GET_SETTINGS.value]
        response = self._call_manager(request)
        return Settings(response[1])

    def get_checkpoint_info(self) -> _CheckpointInfoType:
        """Get the checkpoint info from the manager.

        Returns:
            elapsed_time: current elapsed time
            checkpoints: checkpoint configuration
            resume: path to the resume snapshot
            snapshot_directory: path to store snapshots
        """
        request = [RequestType.GET_CHECKPOINT_INFO.value, str(self._instance_id)]
        response = self._call_manager(request)
        return decode_checkpoint_info(*response[1:])

    def register_instance(
            self, locations: List[str], ports: List[Port]) -> None:
        """Register a component instance with the manager.

        Args:
            locations: List of places where the instance can be
                    reached.
            ports: List of ports of this instance.
        """
        request = [
                RequestType.REGISTER_INSTANCE.value,
                str(self._instance_id), locations,
                [encode_port(p) for p in ports],
                libmuscle.__version__]
        response = self._call_manager(request)
        if response[0] == ResponseType.ERROR.value:
            raise RuntimeError(
                    f'Error registering instance: {response[1]}')

    def request_peers(self) -> Tuple[
            List[Conduit],
            Dict[Reference, List[int]],
            Dict[Reference, List[str]]]:
        """Request connection information about peers.

        This will repeat the request at an exponentially increasing
        query interval at first, until it reaches the interval
        specified by PEER_INTERVAL_MIN and PEER_INTERVAL_MAX. From
        there on, intervals are drawn randomly from that range.

        Returns:
            A tuple containing a list of conduits that this instance is
            attached to, a dictionary of peer dimensions, which is
            indexed by Reference to the peer kernel, and specifies how
            many instances of that kernel there are, and a dictionary
            of peer instance locations, indexed by Reference to a peer
            instance, and containing for each peer instance a list of
            network location strings at which it can be reached.
        """
        sleep_time = 0.1
        start_time = perf_counter()

        request = [RequestType.GET_PEERS.value, str(self._instance_id)]
        response = self._call_manager(request)

        while (response[0] == ResponseType.PENDING.value and
               perf_counter() < start_time + PEER_TIMEOUT and
               sleep_time < PEER_INTERVAL_MIN):
            sleep(sleep_time)
            response = self._call_manager(request)
            sleep_time *= 1.5

        while (response[0] == ResponseType.PENDING.value and
               perf_counter() < start_time + PEER_TIMEOUT):
            sleep(uniform(PEER_INTERVAL_MIN, PEER_INTERVAL_MAX))
            response = self._call_manager(request)

        if response[0] == ResponseType.PENDING.value:
            raise RuntimeError('Timeout waiting for peers to appear')

        if response[0] == ResponseType.ERROR.value:
            raise RuntimeError('Error getting peers from manager: {}'.format(
                    response[1]))

        conduits = [Conduit(snd, recv) for snd, recv in response[1]]

        peer_dimensions = {
                Reference(component): dims
                for component, dims in response[2].items()}

        peer_locations = {
                Reference(instance): locs
                for instance, locs in response[3].items()}

        return conduits, peer_dimensions, peer_locations

    def deregister_instance(self) -> None:
        """Deregister a component instance with the manager.
        """
        request = [
                RequestType.DEREGISTER_INSTANCE.value, str(self._instance_id)]
        response = self._call_manager(request)

        if response[0] == ResponseType.ERROR.value:
            raise RuntimeError('Error deregistering instance: {}'.format(
                    response[1]))

    def waiting_for_receive(
            self, peer_instance_id: Reference, port_name: str, slot: Optional[int]
            ) -> None:
        """Notify the manager that we're waiting to receive a message."""
        request = [
                RequestType.WAITING_FOR_RECEIVE.value,
                str(self._instance_id),
                str(peer_instance_id), port_name, slot]
        self._call_manager(request)

    def waiting_for_receive_done(
            self, peer_instance_id: Reference, port_name: str, slot: Optional[int]
            ) -> None:
        """Notify the manager that we're done waiting to receive a message."""
        request = [
                RequestType.WAITING_FOR_RECEIVE_DONE.value,
                str(self._instance_id),
                str(peer_instance_id), port_name, slot]
        self._call_manager(request)

    def is_deadlocked(self) -> bool:
        """Ask the manager if we're part of a deadlock."""
        request = [
                RequestType.IS_DEADLOCKED.value, str(self._instance_id)]
        response = self._call_manager(request)
        return bool(response[1])

    def _call_manager(self, request: Any) -> Any:
        """Call the manager and do en/decoding.

        Args:
            request: The request to encode and send

        Returns:
            The decoded response
        """
        with self._mutex:
            encoded_request = msgpack.packb(request, use_bin_type=True)
            response, _ = self._transport_client.call(encoded_request)
            return msgpack.unpackb(response, raw=False)

File: libmuscle/cpp/src/libmuscle/mmp_client.cpp
#include "libmuscle/mmp_client.hpp"

#include "libmuscle/data.hpp"
#include "libmuscle/mcp/data_pack.hpp"
#include "libmuscle/mcp/protocol.hpp"
#include "libmuscle/version.h"

#include <chrono>
#include <iterator>
#include <memory>
#include <random>
#include <sstream>
#include <string>
#include <thread>
#include <tuple>
#include <utility>
#include <vector>

#include <msgpack.hpp>
#include "ymmsl/ymmsl.hpp"


using libmuscle::_MUSCLE_IMPL_NS::Data;
using libmuscle::_MUSCLE_IMPL_NS::DataConstRef;
using libmuscle::_MUSCLE_IMPL_NS::mcp::unpack_data;
using libmuscle::_MUSCLE_IMPL_NS::Optional;
using libmuscle::_MUSCLE_IMPL_NS::ProfileEvent;
using libmuscle::_MUSCLE_IMPL_NS::SnapshotMetadata;
using std::chrono::steady_clock;
using ymmsl::Conduit;
using ymmsl::Reference;
using ymmsl::SettingValue;


namespace {
    const std::chrono::milliseconds peer_timeout(600000);
    const int peer_interval_min = 5000;     // milliseconds
    const int peer_interval_max = 10000;    // milliseconds

    std::chrono::milliseconds random_sleep_time() {
        static std::default_random_engine engine;
        static std::uniform_int_distribution<int> dist(
                peer_interval_min, peer_interval_max);

        return std::chrono::milliseconds(dist(engine));
    }

    Data encode_operator(ymmsl::Operator op) {
        switch (op) {
            case ymmsl::Operator::NONE:
                return Data("NONE");
            case ymmsl::Operator::F_INIT:
                return Data("F_INIT");
            case ymmsl::Operator::O_I:
                return Data("O_I");
            case ymmsl::Operator::S:
                return Data("S");
            case ymmsl::Operator::O_F:
                return Data("O_F");
        }
        // can't happen, but silence compiler warning
        throw std::runtime_error("Invalid operator");
    }

    Data encode_port(ymmsl::Port const & port) {
        return Data::list(std::string(port.name), encode_operator(port.oper));
    }

    template <class T>
    Data encode_optional(Optional<T> const & value) {
        Data encoded;
        if (value.is_set())
            encoded = value.get();
        return encoded;
    }

    template <typename T>
    Data encode_vector(std::vector<T> const & value) {
        auto retval = Data::nils(value.size());
        for (std::size_t i = 0u; i < value.size(); ++i)
            retval[i] = value[i];
        return retval;
    }

    Data encode_profile_event(ProfileEvent const & event) {
        if (!event.start_time.is_set() || !event.stop_time.is_set()) {
            throw std::runtime_error(
                    "Incomplete ProfileEvent sent. This is a bug, please"
                    " report it.");
        }

        Data encoded_port;
        if (event.port.is_set())
            encoded_port = encode_port(event.port.get());

        return Data::list(
                static_cast<int>(event.event_type),
                event.start_time.get().nanoseconds,
                event.stop_time.get().nanoseconds,
                encoded_port, encode_optional(event.port_length),
                encode_optional(event.slot), encode_optional(event.message_number),
                encode_optional(event.message_size),
                encode_optional(event.message_timestamp));
    }

    Data encode_snapshot_metadata(SnapshotMetadata const & snapshot_metadata) {
        auto port_message_counts = Data::dict();
        for(auto const & kv : snapshot_metadata.port_message_counts)
            port_message_counts[kv.first] = encode_vector(kv.second);

        auto metadata = Data::dict(
                "triggers", encode_vector(snapshot_metadata.triggers),
                "wallclock_time", snapshot_metadata.wallclock_time,
                "timestamp", snapshot_metadata.timestamp,
                "next_timestamp", encode_optional(snapshot_metadata.next_timestamp),
                "port_message_counts", port_message_counts,
                "is_final_snapshot", snapshot_metadata.is_final_snapshot,
                "snapshot_filename", snapshot_metadata.snapshot_filename
        );

        return metadata;
    }

    template <typename T>
    Optional<T> decode_optional(DataConstRef const & data) {
        if (data.is_nil())
            return {};
        return data.as<T>();
    }
}

namespace libmuscle { namespace _MUSCLE_IMPL_NS {

MMPClient::MMPClient(
        Reference const & instance_id, std::string const & location)
    : instance_id_(instance_id)
    , transport_client_(location)
{}

void MMPClient::close() {
    transport_client_.close();
}

void MMPClient::submit_log_message(LogMessage const & message) {
    auto request = Data::list(
            static_cast<int>(RequestType::submit_log_message),
            message.instance_id,
            message.timestamp.seconds,
            static_cast<int>(message.level),
            message.text);

    call_manager_(request);
}

void MMPClient::submit_profile_events(
        std::vector<ProfileEvent> const & events)
{
    auto event_list = Data::nils(events.size());
    for (std::size_t i = 0u; i < events.size(); ++i)
        event_list[i] = encode_profile_event(events[i]);

    auto request = Data::list(
            static_cast<int>(RequestType::submit_profile_events),
            static_cast<std::string>(instance_id_),
            event_list);

    auto response = call_manager_(request);
}

void MMPClient::submit_snapshot_metadata(
        SnapshotMetadata const & snapshot_metadata) {
    auto request = Data::list(
            static_cast<int>(RequestType::submit_snapshot),
            static_cast<std::string>(instance_id_),
            encode_snapshot_metadata(snapshot_metadata));

    auto response = call_manager_(request);
}

void MMPClient::register_instance(
        std::vector<std::string> const & locations,
        std::vector<::ymmsl::Port> const & ports)
{
    auto encoded_locs = encode_vector(locations);
    auto encoded_ports = Data::nils(ports.size());
    for (std::size_t i = 0u; i < ports.size(); ++i)
        encoded_ports[i] = encode_port(ports[i]);

    auto request = Data::list(
            static_cast<int>(RequestType::register_instance),
            static_cast<std::string>(instance_id_), encoded_locs,
            encoded_ports, MUSCLE3_VERSION);

    auto response = call_manager_(request);

    if (response[0].as<int>() == static_cast<int>(ResponseType::error))
        throw std::runtime_error(
                "Error registering instance: " + response[1].as<std::string>());
}

ymmsl::Settings MMPClient::get_settings() {
    auto request = Data::list(static_cast<int>(RequestType::get_settings));
    auto response = call_manager_(request);
    // always returns success, so no need to check

    auto dict = response[1];

    ymmsl::Settings settings;
    for (std::size_t i = 0u; i < dict.size(); ++i)
        settings[dict.key(i)] = dict.value(i).as<SettingValue>();

    return settings;
}

auto MMPClient::get_checkpoint_info() ->
        std::tuple<
            double,
            DataConstRef,
            Optional<std::string>,
            Optional<std::string>
        >
{
    auto request = Data::list(
            static_cast<int>(RequestType::get_checkpoint_info),
            static_cast<std::string>(instance_id_));
    auto response = call_manager_(request);

    if (response[0].as<int>() != static_cast<int>(ResponseType::success)) {
        throw std::runtime_error("Error getting checkpoint info from manager.");
    }

    return std::make_tuple(
            response[1].as<double>(),
            response[2],
            decode_optional<std::string>(response[3]),
            decode_optional<std::string>(response[4]));
}

auto MMPClient::request_peers() ->
        std::tuple<
            std::vector<::ymmsl::Conduit>,
            std::unordered_map<::ymmsl::Reference, std::vector<int>>,
            std::unordered_map<::ymmsl::Reference, std::vector<std::string>>
        >
{
    int sleep_time = 100;   // milliseconds
    auto start_time = steady_clock::now();

    auto request = Data::list(
            static_cast<int>(RequestType::get_peers),
            static_cast<std::string>(instance_id_));
    auto response = call_manager_(request);

    const int status_pending = static_cast<int>(ResponseType::pending);

    while ((response[0].as<int>() == status_pending) &&
           (steady_clock::now() < start_time + peer_timeout) &&
           (sleep_time < peer_interval_min)) {
        std::this_thread::sleep_for(std::chrono::milliseconds(sleep_time));
        response.reseat(call_manager_(request));
        sleep_time += sleep_time / 2;
    }

    while ((response[0].as<int>() == status_pending) &&
           (steady_clock::now() < start_time + peer_timeout)) {
        std::this_thread::sleep_for(random_sleep_time());
        response.reseat(call_manager_(request));
    }

    if (response[0].as<int>() == status_pending)
        throw std::runtime_error("Timeout waiting for peers to appear.");

    if (response[0].as<int>() == static_cast<int>(ResponseType::error)) {
        std::ostringstream oss;
        oss << "Error getting peers from manager: " << response[1].as<std::string>();
        throw std::runtime_error(oss.str());
    }

    auto const & recv_conduits = response[1];
    std::vector<Conduit> conduits;
    for (std::size_t i = 0u; i < recv_conduits.size(); ++i)
        conduits.emplace_back(
                recv_conduits[i][0].as<std::string>(), recv_conduits[i][1].as<std::string>());


    auto const & recv_dims = response[2];
    std::unordered_map<Reference, std::vector<int>> peer_dimensions;
    for (std::size_t i = 0u; i < recv_dims.size(); ++i) {
        auto dim_list = recv_dims.value(i);
        std::vector<int> dims(dim_list.size());
        for (std::size_t j = 0u; j < dim_list.size(); ++j)
            dims[j] = dim_list[j].as<int>();

        peer_dimensions[recv_dims.key(i)] = dims;
    }

    auto const & recv_locs = response[3];
    std::unordered_map<Reference, std::vector<std::string>> peer_locations;
    for (std::size_t i = 0u; i < recv_locs.size(); ++i) {
        auto peer_locs = recv_locs.value(i);
        std::vector<std::string> locs(peer_locs.size());
        for (std::size_t j = 0u; j < peer_locs.size(); ++j)
            locs[j] = peer_locs[j].as<std::string>();

        peer_locations[recv_locs.key(i)] = locs;
    }

    return std::make_tuple(
            std::move(conduits),
            std::move(peer_dimensions),
            std::move(peer_locations));
}

void MMPClient::deregister_instance() {
    auto request = Data::list(
            static_cast<int>(RequestType::deregister_instance),
            static_cast<std::string>(instance_id_));
    auto response = call_manager_(request);
    if (response[0].as<int>() == static_cast<int>(ResponseType::error)) {
        std::ostringstream oss;
        oss << "Failed to deregister: " << response[1].as<std::string>();
        throw std::runtime_error(oss.str());
    }
}

void MMPClient::waiting_for_receive(
        std::string const & peer_instance_id, std::string const & port_name,
        Optional<int> slot)
{
    auto request = Data::list(
            static_cast<int>(RequestType::waiting_for_receive),
            static_cast<std::string>(instance_id_),
            peer_instance_id, port_name, encode_optional(slot));

    auto response = call_manager_(request);
}

void MMPClient::waiting_for_receive_done(
        std::string const & peer_instance_id, std::string const & port_name,
        Optional<int> slot)
{
    auto request = Data::list(
            static_cast<int>(RequestType::waiting_for_receive_done),
            static_cast<std::string>(instance_id_),
            peer_instance_id, port_name, encode_optional(slot));

    auto response = call_manager_(request);
}

bool MMPClient::is_deadlocked() {
    auto request = Data::list(
            static_cast<int>(RequestType::is_deadlocked),
            static_cast<std::string>(instance_id_));
    
    auto response = call_manager_(request);
    return response[1].as<bool>();
}

DataConstRef MMPClient::call_manager_(DataConstRef const & request) {
    std::lock_guard<std::mutex> lock(mutex_);

    msgpack::sbuffer sbuf;
    msgpack::pack(sbuf, request);

    auto res = transport_client_.call(sbuf.data(), sbuf.size());
    auto const & result = std::get<0>(res);

    auto zone = std::make_shared<msgpack::zone>();
    return unpack_data(zone, result.data(), result.size());
}


} }

